File: sdk-source/lua/examples/quickstart.lua
-- Runnable quickstart for the Aegis Lua SDK.
--
--   AEGIS_APP_KEY=... lua -e "package.path='src/?.lua;src/?/init.lua;'..package.path" examples/quickstart.lua
package.path = 'src/?.lua;src/?/init.lua;' .. package.path

local Aegis = require('aegis')

local function envOr(key, fallback)
  local value = os.getenv(key)
  if value == nil or value == '' then return fallback end
  return value
end

local aegis = Aegis.new({
  baseUrl = envOr('AEGIS_BASE_URL', 'http://localhost:8080'),
  appKey = envOr('AEGIS_APP_KEY', ''),
  version = '1.0.0',
})

local ok, err = pcall(function()
  local info = aegis:init()
  print('initialized: ' .. tostring(info.status))

  if info.version and info.version.update_required then
    print('mandatory update: ' .. tostring(info.version.latest))
    return
  end

  local auth = aegis:login(envOr('AEGIS_USERNAME', 'demo'), envOr('AEGIS_PASSWORD', 'demo-password'))
  print('signed in as ' .. tostring(auth.user and auth.user.username))

  aegis:setVariable('last_seen', os.date('!%Y-%m-%dT%H:%M:%SZ'))
  print('authenticated: ' .. tostring(aegis:isAuthenticated()))

  aegis:heartbeat()
  aegis:logout()
  print('signed out.')
end)

if not ok then
  io.stderr:write(tostring(err) .. '\n')
  os.exit(1)
end

File: sdk-source/java/examples/Quickstart.java
// Runnable quickstart for the Aegis Java SDK.
//
//   mvn -q package
//   java -cp target/aegis-sdk-1.0.0.jar examples/Quickstart.java
import io.aegis.sdk.Aegis;
import io.aegis.sdk.AegisException;
import io.aegis.sdk.AegisOptions;

import java.time.Duration;
import java.util.Map;

public class Quickstart {
    public static void main(String[] args) throws Exception {
        String baseUrl = envOr("AEGIS_BASE_URL", "http://localhost:8080");
        String appKey = envOr("AEGIS_APP_KEY", "");

        try (Aegis aegis = new Aegis(new AegisOptions(baseUrl, appKey).version("1.0.0"))) {
            Map<String, Object> info = aegis.init();
            System.out.println("initialized: " + info.get("status"));

            Object versionInfo = info.get("version");
            if (versionInfo instanceof Map<?, ?> version && Boolean.TRUE.equals(version.get("update_required"))) {
                System.out.println("mandatory update: " + version.get("latest"));
                return;
            }

            Map<String, Object> auth = aegis.login(envOr("AEGIS_USERNAME", "demo"), envOr("AEGIS_PASSWORD", "demo-password"));
            System.out.println("signed in as " + ((Map<?, ?>) auth.get("user")).get("username"));

            System.out.println("user variables: " + aegis.getVariables("user", null).get("variables"));
            aegis.setVariable("last_seen", java.time.Instant.now().toString(), "user", null);

            aegis.startHeartbeat(Duration.ofSeconds(60), reason -> System.out.println("session revoked: " + reason));
            Thread.sleep(2000);

            aegis.logout();
            System.out.println("signed out.");
        } catch (AegisException error) {
            System.err.printf("Aegis error [%s] %s%n", error.getCode(), error.getMessage());
            System.exit(1);
        }
    }

    private static String envOr(String key, String fallback) {
        String value = System.getenv(key);
        return value == null || value.isBlank() ? fallback : value;
    }
}